File: src/main/java/net/igneo/icv/enchantmentActions/enchantManagers/armor/StasisManager.java
package net.igneo.icv.enchantmentActions.enchantManagers.armor;

import net.igneo.icv.client.indicators.EnchantIndicator;
import net.igneo.icv.client.indicators.StasisCooldownIndicator;
import net.igneo.icv.enchantment.EnchantType;
import net.minecraft.util.Mth;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.Vec3;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class StasisManager extends ArmorEnchantManager{
    public StasisManager(Player player) {
        super(EnchantType.BOOTS, 300, -10, true, player);
    }
    public HashMap<Entity,StasisEntityDataManager> entityData = new HashMap<>();
    private List<Entity> blackList = new ArrayList<>();
    @Override
    public void activate() {
        System.out.println("activating");
        for (Entity entity : player.level().getEntities(null,player.getBoundingBox().inflate(20))) {
            if (!(entity instanceof LivingEntity) && !blackList.contains(entity)) {
                entityData.put(entity,new StasisEntityDataManager(Vec3.ZERO,entity.position(),entity.getLookAngle()));
                entity.addTag("stasis");
            }
        }
        active = true;
    }

    @Override
    public void onOffCoolDown(Player player) {

    }

    @Override
    public EnchantIndicator getIndicator() {
        return new StasisCooldownIndicator(this);
    }

    @Override
    public boolean canUse() {
        return !active;
    }

    @Override
    public void dualActivate() {
        release();
        resetCoolDown();
        System.out.println("no more active");
    }

    @Override
    public boolean isDualUse() {
        return true;
    }

    public void release() {
        active = false;
        for (Entity entity : entityData.keySet()) {
            entity.removeTag("stasis");
            entity.setDeltaMovement(entityData.get(entity).getMovement());
        }
        entityData = new HashMap<>();
    }

    @Override
    public void tick() {
        super.tick();
        if (active) {
            ++activeTicks;
            for (Entity entity : entityData.keySet()) {
                entity.setPos(entityData.get(entity).position);
                entityData.get(entity).addMovement(entity.getDeltaMovement());
                entity.setDeltaMovement(Vec3.ZERO);
                float yaw = (float) (Mth.atan2(-entityData.get(entity).look.x, entityData.get(entity).look.z) * (180.0F / Math.PI));
                float pitch = (float) (Mth.atan2(-entityData.get(entity).look.y, entityData.get(entity).look.horizontalDistance()) * (180.0F / Math.PI));

                // Apply new look angles
                entity.setYRot(yaw);
                entity.setXRot(pitch);
            }
        } else {
            activeTicks = 0;
        }
        if (activeTicks > 1200) {
            dualActivate();
        }
    }

    public void addMovement(Entity entity, Vec3 movement) {
        entityData.get(entity).addMovement(movement);
    }

    class StasisEntityDataManager {
        private Vec3 movement;
        public Vec3 position;
        public Vec3 look;
        StasisEntityDataManager(Vec3 movement, Vec3 position, Vec3 look) {
            this.movement = new Vec3(movement.x,0, movement.z);
            addMovement(Vec3.ZERO);
            this.position = position;
            this.look = look;
        }
        public Vec3 getMovement() {
            return movement;
        }
        public void addMovement(Vec3 push) {
            double max = 6;
            double d0 = Math.max(-max, Math.min(movement.x + push.x, max));
            double d1 = Math.max(0, Math.min(movement.y + push.y, 3));
            double d2 = Math.max(-max, Math.min(movement.z + push.z, max));

            this.movement = new Vec3(d0,d1,d2);
        }
    }
}

File: src/main/java/net/igneo/icv/enchantmentActions/enchantManagers/armor/boots/StasisManager.kt
package net.igneo.icv.enchantmentActions.enchantManagers.armor.boots

import net.igneo.icv.client.indicators.EnchantIndicator
import net.igneo.icv.client.indicators.StasisCooldownIndicator
import net.igneo.icv.enchantment.EnchantType
import net.igneo.icv.enchantmentActions.enchantManagers.armor.ArmorEnchantManager
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.phys.Vec3
import kotlin.math.max
import kotlin.math.min

class StasisManager(player: Player?) :
    ArmorEnchantManager(EnchantType.BOOTS, 300, -10, true, player) {
    @JvmField
    var entityData: HashMap<Entity, StasisEntityDataManager> = HashMap()
    private val blackList: List<Entity> = ArrayList()

    override fun activate() {
        println("activating")
        for (entity in player!!.level().getEntities(null, player!!.boundingBox.inflate(20.0))) {
            if (entity !is LivingEntity && !blackList.contains(entity)) {
                entityData[entity] = StasisEntityDataManager(Vec3.ZERO, entity.position(), entity.yRot, entity.xRot)
            }
        }
        active = true
    }

    override fun onOffCoolDown(player: Player?) {
    }

    override val indicator: EnchantIndicator
        get() = StasisCooldownIndicator(this)

    override fun canUse(): Boolean {
        return !active
    }

    override fun dualActivate() {
        release()
        resetCoolDown()
        println("no more active")
    }

    override var isDualUse: Boolean
        get() = true
        set(isDualUse) {
            super.isDualUse = isDualUse
        }

    fun release() {
        active = false
        for (entity in entityData.keys) {
            entity.removeTag("stasis")
            entity.deltaMovement = entityData[entity]!!.movement
        }
        entityData = HashMap()
    }

    override fun tick() {
        super.tick()
        if (active) {
            ++activeTicks
            for (entity in entityData.keys) {
                entity.setPos(entityData[entity]!!.position)
                entityData[entity]!!.addMovement(entity.deltaMovement)
                entity.deltaMovement = Vec3.ZERO
                val yaw = entityData[entity]!!.YRot
                val pitch = entityData[entity]!!.XRot


                // Apply new look angles
                entity.yRot = yaw
                entity.xRot = pitch
            }
        } else {
            activeTicks = 0
        }
        if (activeTicks > 1200) {
            dualActivate()
        }
    }

    fun addMovement(entity: Entity, movement: Vec3) {
        entityData[entity]!!.addMovement(movement)
    }

    inner class StasisEntityDataManager(movement: Vec3, position: Vec3, YRot: Float, XRot: Float) {
        var movement: Vec3
            private set
        var position: Vec3
        var YRot: Float
        var XRot: Float

        init {
            this.movement = Vec3(movement.x, 0.0, movement.z)
            addMovement(Vec3.ZERO)
            this.position = position
            this.YRot = YRot
            this.XRot = XRot
        }

        fun addMovement(push: Vec3) {
            val max = 6.0
            val d0 = max(-max, min(movement.x + push.x, max))
            val d1 = max(0.0, min(movement.y + push.y, 3.0))
            val d2 = max(-max, min(movement.z + push.z, max))

            this.movement = Vec3(d0, d1, d2)
        }
    }
}


